Add debug logging

araglu opened the pull request:
CHW-428
This change adds logging for several UIT+ calls:
- For get_userinfo(), it adds only an INFO log line
- For call(), get_file(), put_file(), and list_directory(), it adds one INFO line and many DEBUG lines.
  - The INFO line for call() will show the shell command in cyan text and working directory
  - The INFO line for get_file() and put_file() will show the remote file and local file
  - The INFO line for list_directory() will show the remote directory
  - The DEBUG lines for these four functions will show:
    - Duration of the request in red text
    - Login node
    - Return code of the shell command
    - HPCMP username
    - if they are not empty, stdout and stderr will show up to 500 characters in each, and all line breaks are shown as \n
    - A brief stacktrace that only shows our codebases and ignores python standard library, bokeh, panel, param, django, tethys, etc.

Most of the debug code will not be executed unless logging has DEBUG enabled. I had one instance where it tried to send the unicode icons in the File Browser to logging, and the logging module crashed on it. I found a workaround for that specific case, but I never want to see anything like that happen on a prod server, especially if the prod server is not logging this information. 

Here is an example of the debug output:
![uit-debug](https://user-images.githubusercontent.com/14322382/224152747-04d8bd7a-8404-45fd-ae39-5ece257e926c.png)
To get pyuit debug logging, you will need to modify your ~/.tethys/<conda_env>/portal_config.yml with at least this much yaml:
```
settings:
  LOGGING:
    loggers:
      uit:
        handlers:
          - console_verbose
        level: DEBUG 
```
If you want less logging, set the level from DEBUG to INFO.

This only shows stacktrace lines from 'pyuit' be default, which only works in a dev environment. To show lines from your codebases, add these lines to a config file named ~/.uit:
```
debug_stacktrace_allowlist:
  - uit
  - your_codebase_dir
```
To see debug logging in Jupyter Lab, run this code to show debug output:
```
import logging
handler = logging.StreamHandler()
formatter = logging.Formatter('%(asctime)s %(levelname)s:%(name)s:%(message)s')
handler.setFormatter(formatter)
logger = logging.getLogger('uit')
logger.addHandler(handler)
logger.setLevel('DEBUG') 
```

araglu committed:
Add detailed logs for UIT+ calls

Show the command and working directory in logger.INFO, and then show the
execution time of the command, login node, return code, stdout, stderr,
and a brief stack trace in logger.DEBUG.

araglu committed:
Add username and fix unicode string

- Add username to first line of debug output
- Removes the b'' text around the stacktrace function call
- code cleanup

araglu committed:
Fix duplicate stacktrace lines

araglu committed:
Fix E126 continuation line over-indented for hanging indent

But not fixing "E127 continuation line over-indented for visual indent"
because I don't see the problem there.

araglu committed:
Improve json error handling

This used to error out on every get_file()

araglu committed:
Ignore qdel warning (#37)

* Remove working_dir from _execute()
qdel would fail when run against an archived job when its main job was already deleted.

* Handle qdel rc=35 when ran against a finished job

* Support qdel 'Unknown Job Id' result

araglu committed:
File Browser better symlink handling and speed improvement (#38)

* Remove working_dir from _execute()
qdel would fail when run against an archived job when its main job was already deleted.

* Handle qdel rc=35 when ran against a finished job

* Determine if symlink is dir or file, and stop running 'ls -l /p/home/'
It can now descend into a symlinked directory, but this still does not work well if there is a broken symlink in the directory.
Changed it to run 'ls -ld' on the current directory instead of 'ls -l' on the parent. That was taking 10 seconds on the user's home directory on an HPC with 7000 users.

* Dig deep into symlinks and handle broken symlinks
This now doesn't hide all symlinks if one is broken, but it is very slow, and it doesn't show files.

* Revert "Dig deep into symlinks and handle broken symlinks"

This reverts commit 1657919d42ca7644dc94096fb18a28072f43f420.

* Change 'ls -ld' to 'ls -ldL' so it resolves symlinks itself
This code and UIT+ only care about dirs and files, so symlinks only get in the way. This will quickly show target files while ignoring broken symlinks.
Not to be confused with bad cholesterol.

* Removed duplicate code

* Initialize HpcPath by / operator

* Removing commits from a different PR

---------

Co-authored-by: sdc50 <sdc50@byu.net>

sdc50 commented on uit/uit.py:
```diff
@@ -726,6 +745,92 @@ def _parse_hpc_output(cls, output, as_df, columns=None, delimiter=None, delimite
+                nice_stderr += "'" + resp.get('stderr').replace('\n', '\\n') + "'"
+
+        # Show only relevant function calls and ignore standard library
+        only_show_stacktrace = ['pyuit',
```
@araglu PyUIT is designed to be a generic library and is used by other projects, so we shouldn't have any project specific references in there (e.g. helios, tethys, uit_plus, etc.). Could we make this list configurable? Maybe the default is just `['pyuit']`, but then it can check the `self.config_file` for a user specified list to use instead.

araglu replied:
I changed it to grab a list from the config_file.
Thanks for the suggestion! I wanted to do something that would only show dev code automagically such as filtering out site-packages, but I doubt that would work on a dev server, and I may want to peek inside one or two libraries occasionally.

sdc50 commented on uit/uit.py:
```diff
@@ -726,6 +744,92 @@ def _parse_hpc_output(cls, output, as_df, columns=None, delimiter=None, delimite
+                            trimmed_filename = os.sep.join(stacktrace[i].filename.split(os.sep)[j:])
+                            break
+                    nice_trace += f"\n    {i}: {trimmed_filename}:" + \
+                                  f"{stacktrace[i].lineno} {stacktrace[i].name}()" + \
```
I'm not overly concerned about the stickler errors here, but since you are making some changes anyway, maybe consider something like this?

```python
 nice_trace += (
    f"\n    {i}: {trimmed_filename}:"
    f"{stacktrace[i].lineno} {stacktrace[i].name}()"
    f"    {stacktrace[i].line.encode('ascii', 'backslashreplace').decode('ascii')}"
)
```

araglu replied:
I learned that PEP8 prefers your code style. Fixed.

sdc50 reviewed (changes requested):
@araglu Thanks for this contribution. I think it will make pyuit nicer to use/debug. I made a few minor suggestions, primarily to keep pyuit a generic library.

araglu committed:
Retrieve debug stacktrace options from config file

- Load debug_stacktrace_allowlist from PyUIT yaml config file
- Always attempt to load the config file now, instead of only loading if
there is no UIT token or ID
- Condense the stdout and stderr code
- Change the regex to a raw string to avoid lint warnings
- Fix a typo in README.rst
- Remove quotes in tox.ini to get local flake8 to run correctly
